Implement Unicode Name - NR2

eyeplum opened the pull request:
This is a work in progress for implementing Unicode Name - NR2 (#171).

- [x] NR2: CJK Unified Ideograph / Tangut Ideograph / Nushu Character / CJK Compatibility Ideograph
- [x] Add tests

Feedback is welcome!

eyeplum commented:
Thanks for the review!
I will update according to the comments and add some tests, also will put off `NR1` for now.
I already have some rough ideas about how `NR1` could be implemented, so hopefully a new PR about that will come soon :)
Thanks again.

eyeplum commented on unic/ucd/name/src/name.rs:
```diff
@@ -10,37 +10,138 @@
+                Ok(())
+            },
+            &Name::HangulSyllable(ch) => {
+                // FIXME: unimplemented
```
Added a temporary implementation with `HANGUL SYLLABLE {HEX_VALUE}`.

eyeplum commented on unic/ucd/name/src/name.rs:
```diff
@@ -10,37 +10,138 @@
+        }
+    }
+
+    fn num_of_digits(&self, code_point: u32) -> usize {
```
Fixed.

eyeplum commented on unic/ucd/name/src/name.rs:
```diff
@@ -10,37 +10,138 @@
     pub fn of(ch: char) -> Option<Name> {
-        data::NAMES.find(ch).map(|pieces| Name { pieces })
+        let code_point = ch as u32;
+        match code_point {
```
Fixed.

eyeplum commented on unic/ucd/name/src/name.rs:
```diff
@@ -10,37 +10,138 @@
+pub static PREFIX_NUSHU_CHARACTER: &'static str = "NUSHU CHARACTER-";
+pub static PREFIX_CJK_COMPATIBILITY_IDEOGRAPH: &'static str = "CJK COMPATIBILITY IDEOGRAPH-";
+
 #[derive(Copy, Clone, Debug, Ord, PartialOrd, Eq, PartialEq, Hash)]
```
Ended up with a big 2-level matching, happy to change it if there is a more concise way.

behnam commented on unic/ucd/name/tests/basic_tests.rs:
```diff
@@ -13,12 +13,37 @@ extern crate unic_ucd_name;
-fn basic_tests() {
+fn test_name_str() {
     assert_eq!(
         Name::of('A').expect("No name for A").to_string(),
```
nit: In tests, it's better to not use `expect()`, unless it's something that may fail. I think using `unwrap()` in these cases make the test more readable by being shorter, and more clean about expectation.

eyeplum replied:
Sounds good, fixed :)

behnam commented on unic/ucd/name/tests/basic_tests.rs:
```diff
@@ -13,12 +13,37 @@ extern crate unic_ucd_name;
+#[test]
+fn test_name_cmp() {
     assert!(Name::of('A') < Name::of('B'));
+    assert!(Name::of('言') < Name::of('A'));
```
Would you please expand the test here to compare `Name::of()` AND their `to_string()` for any pair of sample?

eyeplum replied:
Expanded the test with more comparison between samples and their `to_string()` comparison.

behnam commented on unic/ucd/name/tests/basic_tests.rs:
```diff
@@ -13,12 +13,37 @@ extern crate unic_ucd_name;
+        Name::of('A').unwrap().to_string(),
         "LATIN CAPITAL LETTER A"
     );
+    assert_eq!(
```
Uh, another nit: It's good practice to group the asserts together based on their rule in the spec, and have a comment for each group with the rule name. Like this:
```
// == NR1: Hangul Jamos ==

assert_eq!(
    ...
)
assert_eq!(
    ...
)

// == NR2: Ideographics ==

assert_eq!(
    ...
)
assert_eq!(
    ...
)

// == NR3: ... ==

assert_eq!(
    ...
)
assert_eq!(
    ...
)
```

eyeplum replied:
Fixed.

eyeplum commented:
Thanks for the review @behnam!
I have addressed all the comments and rebased to the latest `master`.
I have also run `cargo fmt` and committed the changes.
Although I'm not able to get `cargo clippy` to work - tried to install with `cargo +nightly install clippy`, got compile errors :(

behnam commented on unic/ucd/name/tests/basic_tests.rs:
```diff
@@ -13,12 +13,170 @@ extern crate unic_ucd_name;
+
+    // FIXME: use decomposed jamo short names instead
+    assert!(Name::of('곲') > Name::of('가'));
+    assert!(Name::of('곲').unwrap().to_string() > Name::of('가').unwrap().to_string());
```
Feel like writing a little macro here for the test logic, and only call the macro on test values?

Here are two examples of doing that in our tests:

https://github.com/behnam/rust-unic/blob/c2ed9f869964192f96ac137fc87847aa7cbb1ca5/unic/segment/tests/words_conformance_tests.rs#L28-L39

https://github.com/behnam/rust-unic/blob/c2ed9f869964192f96ac137fc87847aa7cbb1ca5/unic/ucd/common/tests/consistency_tests.rs#L20-L39

eyeplum replied:
Fixed :)

eyeplum commented:
Thanks @behnam for the advice. 
A macro indeed makes the tests look better and easier to change.
I have also squashed a few commits.

CAD97 commented:
(delegate means you have permission to `bors: r+` it now, when you're ready)

CAD97 commented:
bors: delegate+

eyeplum commented:
@CAD97 sure, I will explore and see.
One other thing I'm thinking about is: since `len()` `fmt()` and `cmp()` are all based on the string value of `self`, is it also possible to simply create a `string_value()` method and call it from those methods? Not sure about the memory footprint though.
I notice the initial implementation was going this way but 1a50398 changed it.

CAD97 commented:
The trick is to avoid creating the `String` if unnecessary. Because `Display` gives `ToString` for free, the use-site can create a `String` that way if they want it.

Also note that this crate is `#[no_std]`. We want to keep these libraries available to no-std environments as much as possible.

eyeplum commented:
Ah, make sense. Thanks!

behnam commented on unic/ucd/name/src/name.rs:
```diff
@@ -9,38 +9,234 @@
+        }
+    }
+
+    fn num_of_hex_digits(&self, code_point: u32) -> usize {
```
You can write this as:

```rust
fn number_of_hex_digits(ch: char) -> usize {
    (32 - u32::leading_zeros(ch as u32) as usize + 3) / 4
}
```

Please note that it should be a static method, since `&self` is not needed in the impl. (`clippy` would have caught this, I think.)

CAD97 replied:
(I knew there existed a clever solution but I couldn't think of it)

behnam replied:
You'll get to master leading-zero/trailing-zero kinds of questions after school, in job interviews! ;)

behnam replied:
Oh, btw, @eyeplum, either make this a macro, or `#[inline=always]` it. There's no reason for it not being inline.

behnam replied:
Here's the `libcore` implementation of (still unstable) `char::escape_unicode`, which uses the same trick, but a bit more readable:

https://github.com/rust-lang/rust/blob/ddaab61101e48f0a58ce176b5a6aff11ae36c986/src/libcore/char.rs#L416-L432

For more fun like this, see this list: https://github.com/rust-lang/rust/search?utf8=%E2%9C%93&q=leading_zeros&type=

And, deep down, this is the impl: https://github.com/rust-lang/rust/blob/a538fe7ce715c7bd27e2e05329c3d857b9ad92af/src/libcore/num/mod.rs#L1357

Which calls LLVM's `ctlz`: https://llvm.org/docs/LangRef.html#llvm-ctlz-intrinsic

eyeplum replied:
Fixed, thank you!

eyeplum commented:
@CAD97 Tried with `enum Name { NR1, NR2, NR3 }`, it does simplify things a lot! So let do it this way!
Thanks for the advice!

eyeplum commented:
bors: r+

behnam commented on unic/ucd/name/src/name.rs:
```diff
@@ -9,38 +9,138 @@
+pub static PREFIX_CJK_COMPATIBILITY_IDEOGRAPH: &'static str = "CJK COMPATIBILITY IDEOGRAPH-";
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum Name {
```
This looks more clean now! Thanks, @eyeplum, for the update.

If you like working more on this component, would be great to add some documentation here with links to the specification of NR1, etc.

Here are two examples:

https://github.com/behnam/rust-unic/blob/master/unic/ucd/age/src/age.rs#L16-L39

https://github.com/behnam/rust-unic/blob/master/unic/ucd/bidi/src/bidi_class.rs#L17-L23

Also, you can add
```rust
#![deny(missing_docs)]
```
to `src/lib.rs` of this component, to require documentation on all public items on compile time.

eyeplum replied:
Sure! I will be working on `NR1` soon and will add the documentations.
Thank you so much for your guidance!
